### PR TITLE
[Kernel] Add exception principles for Kernel to solidify the rules for our exception framework

### PR DESCRIPTION
**Pull request by allisonport-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### CODE REVIEW

File: kernel/EXCEPTION_PRINCIPLES.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+- The input data violates table constraints when writing to the table.
+- Kernel doesn’t support reading a table with XXX table feature.
+
+Developer-facing exceptions:
+- `getInt` is called on a boolean `ColumnVector`.
+- A column mapping mode besides “none”, “id”, and “name” is encountered.
+- An empty iterator is returned from the `Engine` implementation when reading files.
+
+### Wrapping exceptions thrown from the Engine implementation
+We want to wrap any unchecked exceptions thrown from the `Engine` implementation with `KernelEngineException` and include additional context about the failing operation. This makes it clear where the exception is originating from, and the additional context can help future debugging.
```
Comment (vkorukanti) [inline kernel/EXCEPTION_PRINCIPLES.md:37]: Do we want to mention/clarify.
1) Sometimes, the checked exceptions can't be thrown from the `CloseableIterator,`, and instead, we wrap the checked exception in a `KernelEngineException` and throw.
2) Criteria for choosing a checked exception in Engine API specification. It is highly unlikely the exception gets thrown within the class and will be thrown only when the returned data is accessed

**Review by vkorukanti — COMMENTED** [review]
LGTM.

I think we may need to think through how to handle the exceptions happening the lazy iterators. Should we pass some context to them when creating the iterators? This can be a separate task, need not block this PR.

File: kernel/EXCEPTION_PRINCIPLES.md (hunk 1 above)
Comment (allisonport-db) [reply kernel/EXCEPTION_PRINCIPLES.md:37]: Clarified (1). I think (2) is just something we should consider when reviewing interfaces/making design decisions.

**Comment by allisonport-db** [comment]
> I think we may need to think through how to handle the exceptions happening the lazy iterators. Should we pass some context to them when creating the iterators? This can be a separate task, need not block this PR.

I think this would be _nice to have_ but might be a non-trivial design/code change (not sure how we'd incorporate this?) for a smaller gain in debuggability. I'm viewing most of these exception changes as best effort to improve our error messages... this seems more complex than we need right now.

**Review by vkorukanti — APPROVED** [review]

**Pull request merged**